### PR TITLE
Feature/native source maps in @loopback/build

### PR DESCRIPTION
**Pull request by itamar244**
- Added `.mocharc.js` for `packages/build`, for dynamic mocha config.
- Used `enable-source-maps` flag depending on node version.
- in the beginning Left `.mocharc.json` untouched in `packages/build` because it's being copied to created projects, and tests and users depend on it (it's my guess here, need your approval 😃)
- Later changed to `.mocharc.js` in the dependent tests and scripts and remove `.mocharc.json`

Implements #6393 

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/build/config/.mocharc.json
Comment (raymondfeng) [inline packages/build/config/.mocharc.json:8]: Let's remove `.mocharc.json` to favor `.mocharc.js` and avoid confusions.

**Comment by raymondfeng** [comment]
@itamar244 Thank you for the contribution. I wonder if we should improve https://github.com/strongloop/loopback-next/tree/master/packages/cli/generators/project/templates as well. The templates control code generation by `lb4` command.

File: packages/build/config/.mocharc.json
Comment (itamar244) [reply packages/build/config/.mocharc.json:8]: I tried but it fails [here](https://github.com/strongloop/loopback-next/blob/master/packages/build/test/integration/scripts.integration.js#L315).

I think these tests check if `.mocharc.json` is copied to a new generated project, so my guess it's copied to new project.
I can changes these tests to check if `.mocharc.js` is copied correctly, but if I'm right than this will impact users because the generated mocha config file will be different.

I might be completely wrong, need your help here 😄

Comment (itamar244) [reply packages/build/config/.mocharc.json:8]: I think [packages/build/bin/run-mocha.js](https://github.com/strongloop/loopback-next/blob/master/packages/build/bin/run-mocha.js#L37) uses the `.mocharc.json` as default mocha configuration file when running `lb-mocha`.

Maybe if I change it here and in the test in the previous link it will work

Comment (raymondfeng) [reply packages/build/config/.mocharc.json:8]: We can replace all occurrences of `.mocharc.json` with `.mocharc.js`.

File: packages/build/bin/run-mocha.js
Comment (raymondfeng) [inline packages/build/bin/run-mocha.js:37]: Let's use the following to keep backward compatibility:

```js
const mochaOptsFile = utils.getConfigFile('.mocharc.js', '.mocharc.json`);
```

**Review by raymondfeng — APPROVED** [review]
Let's squash all commits into one. Thanks!

**Review by bajtos — APPROVED** [review]
Code changes LGTM, thank you @itamar244 for the pull request!

I'd like to discuss the impact on existing project. Based on the changes you had to make in the top-level `.mocharc.js`, I think this is a breaking change, because the following code will stop working with the new `@loopback/build` version:

```ts
// in my custom project - file .mocharc.json

const defaultConfig = require('@loopback/build/config/.mocharc.json');

module.exports = {
  ...defaultConfig,
  timeout: 5000,
};
```

Please see [Making breaking changes](https://loopback.io/doc/en/lb4/code-contrib-lb4.html#making-breaking-changes) for guidelines for this situation. The most important part is to add `BREAKING CHANGE:` trailer to the commit message and clearly explain what's changing and how can affected users upgrade their projects.

**Review by bajtos — CHANGES_REQUESTED** [review]
On the second thought, please rework [packages/cli/generators/project/templates/.mocharc.json](https://github.com/strongloop/loopback-next/blob/73235e43316ea48e78b74eb42e0a01266bfab7ac/packages/cli/generators/project/templates/.mocharc.json) and apply the same changes. This is the file we are using when scaffolding a new LoopBack project.

**Comment by raymondfeng** [comment]
> On the second thought, please rework packages/cli/generators/project/templates/.mocharc.json and apply the same changes. This is the file we are using when scaffolding a new LoopBack project.

Yes, rename it to `.mocharc.js` and apply similar changes.

File: packages/cli/generators/project/templates/.mocharc.js
Comment (raymondfeng) [inline packages/cli/generators/project/templates/.mocharc.js:1]: Please note this code is going to be run with generated LoopBack applications. The templates for `package.json` does not have `semver` in `devDependencies` yet. It will be problematic.

We also use `source-map-support` in two other places:
- https://github.com/strongloop/loopback-next/blob/master/packages/cli/generators/project/templates/package.json.ejs#L101
- https://github.com/strongloop/loopback-next/blob/master/packages/cli/generators/project/templates/package.json.ejs#L147

Ideally, we need to make them conditional based on the node.js version too. 

Maybe it's not ready for us to update the code generation at this moment for the following issues I'm seeing:
1. A generated LB app can be possibly run with Node 10/12/14. Using `source-map-support` for now is probably the best.
2. It will be difficult to assert snapshots for the generated code which are different between node.js versions.

Comment (itamar244) [reply packages/cli/generators/project/templates/.mocharc.js:1]: so you prefer to wait for node v12 to be the minimum version of loopback and not conditionally polyfill this?

Comment (raymondfeng) [reply packages/cli/generators/project/templates/.mocharc.js:1]: Personally, I would prefer to wait for Node.js 12. I just feel it's not worth the effort that creates a complex solution with not much gain. I'll let @bajtos  to weigh in.

Comment (bajtos) [reply packages/cli/generators/project/templates/.mocharc.js:1]: I don't have a strong opinion. On one hand, I see the benefits of using built-in source map support - for example, it prints information about both TS and JS line location, which makes it easier to troubleshoot errors caused by specific output of TypeScript compiler. On the other hand, I also agree the currently proposed solution is adding quite a lot of complexity, which may not justify the benefits.

Comment (raymondfeng) [reply packages/cli/generators/project/templates/.mocharc.js:1]: Maybe the best compromise is to clean up the PR to use Node.js built-in source map support and wait until Node 10 is EOL to land it.

**Comment by bajtos** [comment]
Not stale, we are waiting for Node.js 10.x to reach end of life in April 2021.

**Comment by achrinza** [comment]
Node.js v10 has reached end-of-life.

**Pull request closed without merging**
